ts/image-occlusion/generate.ts

```diff
@@ -40,22 +37,34 @@ export function generate(hideInactive: boolean): { occlusionCloze: string; noteC
         const obJson = object.toJSON();
         noteCount++;
         if (obJson.type === "group") {
-            clozeData += getGroupCloze(object, index, hideInactive);
+            clozeData += getGroupCloze(canvas, object, index, hideInactive);
         } else {
-            clozeData += getCloze(object, index, null, hideInactive);
+            clozeData += getCloze(canvas, object, index, null, hideInactive);
         }
     });
 
     occlusionCloze += clozeData;
     return { occlusionCloze, noteCount };
 }
 
-const getCloze = (object, index, relativePos, hideInactive): string => {
+function floatToDisplay(number: number): string {
+    return number.toFixed(6).replace(/^0+|0+$/, "");
+}
+
+function xRatio(canvas: HTMLCanvasElement, x: number): string {
```

Comment:
Incredibly minor and maybe not an actual improvement, but from a testing & DRY standpoint, would it perhaps make sense to combine the ratio functions for x and y into one (both here and when restoring to absolute values) and passing in the canvas width/height directly? I.e. a single `toPixelRatio` and `fromPixelRatio` function?

Reply:
I may be missing something, but I'm not sure how this would help. To go from pixels-within-canvas to ratio-within-canvas, we need to divide by either width or height depending on the axis. And same for the other direction - we need to know which axis we're operating on to determine the appropriate pixel.

Reply:
Sorry, I think my comment was a bit ambiguous. I was mostly musing about whether it would maybe make sense to pass the width or height in directly into a couple of common relative <-> absolute conversion functions, i.e. along the lines of:

```ts
function toAxisRatio(part: number, base: number): string {
     return floatToDisplay(part / base);
}

function fromAxisRatio(ratio: string, base: number): string {
     return parseFloat(ratio) * base;
}

// and then e.g. 

clozeData += `:width=${toAxisRatio(obJson.width, canvas.width)}`;

const width = fromAxisRatio(cloze.dataset.width!, canvas.width);
```

Mostly coming out of a place of playing around with the functions in the console before trying the PR and thinking how, if we add tests for them, passing the width & height in directly could simplify things vs. constructing canvas inputs.

But yeah, a tiny improvement if any at all and I feel a bit bad about wasting your time on this

Reply:
I don't have particularly strong feelings here, but given that the current implementation saves about 5-6 characters per line, I slightly prefer it over passing in explicit width/height each time.